### DISCUSSION TITLE
Unknown type name '__int128_t' clang(unknown_typename)

### DISCUSSION THREAD

**Discussion by dmalyuta** [clangd/clangd#2062] [Q&A] [answered]
I'm using clangd through the [VS Code extension](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd). Here is my `compile_commands.json`:
```
[
  {
    "directory": "C:/Users/.../",
    "command": "\"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/Llvm/bin/clang++.exe\" -x c++ \"C:/Users/.../main.cpp\" -std=c++17 -stdlib=libc++ -march=skylake -Wall -fms-compatibility-version=19.10 -Wmicrosoft -Wno-invalid-token-paste -Wno-unknown-pragmas -Wno-unused-value -fsyntax-only \"-DUNICODE\" \"-D_UNICODE\" \"-D_MT\" \"-D_DLL\" \"-DNDEBUG\" \"-D_CONSOLE\" \"-D_DEBUG_FUNCTIONAL_MACHINERY\" -isystem\"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.39.33519/include\" -isystem\"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.39.33519/atlmfc/include\" -isystem\"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Auxiliary/VS/include\" -isystem\"C:/Program Files (x86)/Windows Kits/10/Include/10.0.22621.0/ucrt\" -isystem\"C:/Program Files (x86)/Windows Kits/10/Include/10.0.22621.0/um\" -isystem\"C:/Program Files (x86)/Windows Kits/10/Include/10.0.22621.0/shared\" -isystem\"C:/Program Files (x86)/Windows Kits/10/Include/10.0.22621.0/winrt\" -isystem\"C:/Program Files (x86)/Windows Kits/10/Include/10.0.22621.0/cppwinrt\"",
    "file": "C:/Users/.../main.cpp"
  }
]
```
Most of the file is parsed correctly, however I am unable to get clangd to recognize clang's builtin `__uint128_t` and `__int128_t` 128-bit integer types. Visual Studio IDE has no issues compiling the code, and does recognize them.
![image](https://github.com/clangd/clangd/assets/12475811/743d8176-5cc4-4476-9685-03665236a515)

Any help on resolving this would be very appreciated!

**Comment by HighCommander4** [comment]
Are you able to compile the code with clang as the compiler?

**Reply by dmalyuta** [reply]
Yes, like I said Visual Studio IDE has no problem compiling the code, and also doesn't mark the type as unknown.

**Reply by HighCommander4** [reply]
Yes but does the Visual Studio IDE use clang as its compiler, or MSVC?

**Reply by dmalyuta** [reply]
It uses clang

**Reply by dmalyuta** [reply]
Here is a screenshot of the VS config:
![image](https://github.com/clangd/clangd/assets/12475811/de51f7e3-e2b7-4470-bdb5-d10783580c5a)

![image](https://github.com/clangd/clangd/assets/12475811/19e11a02-274b-4d53-a75b-269435f69a29)

Intellisense in VS picks up the type correctly:
![image](https://github.com/clangd/clangd/assets/12475811/0db72988-b498-422f-b588-feeaeacf5467)

**Comment by dmalyuta** [comment, marked as answer]
Got it fixed by switching to the following `compile_commands.json`:
```json
[
  {
    "directory": "C:/Users/.../",
    "command": "\"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/Llvm/x64/bin/clang-cl.exe\" /c /Z7 /nologo /W3 /WX- /diagnostics:column /O2 /Oi /D __x86_64__ /D __BMI__ /EHsc /MDd /GS /Gy /arch:AVX2 /fp:precise /std:c++17 /Gd /TP -m64 \"C:/Users/.../main.cpp\" -isystem\"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/Llvm/x64/lib/clang/17/include\" -isystem\"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.39.33519/include\" -isystem\"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.39.33519/atlmfc/include\" -isystem\"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Auxiliary/VS/include\" -isystem\"C:/Program Files (x86)/Windows Kits/10/Include/10.0.22621.0/ucrt\" -isystem\"C:/Program Files (x86)/Windows Kits/10/Include/10.0.22621.0/um\" -isystem\"C:/Program Files (x86)/Windows Kits/10/Include/10.0.22621.0/shared\" -isystem\"C:/Program Files (x86)/Windows Kits/10/Include/10.0.22621.0/winrt\" -isystem\"C:/Program Files (x86)/Windows Kits/10/Include/10.0.22621.0/cppwinrt\" -isystem\"C:/Program Files (x86)/Windows Kits/NETFXSDK/4.8/Include/um\"",
    "file": "C:/Users/.../main.cpp"
  }
]
```

TL;DR looks like the [Clang Power Tools](https://clangpowertools.com/) extension for VS IDE does not generate the same compilation command as what VS actually builds with. When I manually copy-paste the right compilate command (as shown above), clangd picks up the type:
![image](https://github.com/clangd/clangd/assets/12475811/65bb44a8-1110-463f-adc4-16aedc73f0ea)
